### DISCUSSION TITLE
Process instance migration?

### DISCUSSION THREAD

**Discussion by Dmytrych** [bpmnServer/bpmn-server#199] [Q&A]
Hi,

Am I correct, that the engine doesn't have a process instance migration yet? Do you have any thoughts about how it can be added into the application?

I am currently considering this engine for my personal project, but I am afraid that the long-running processes will be stuck without migration functionality.

**Comment by ralphhanna** [comment]
Hi @Dmytrych,
What do you mean by `migration`? currently, there is an instance recovery procedure https://bpmnserver.github.io/docs/advanced/#hung-tasks.
You add your own script to detect hung processes and recover them.

**Reply by Dmytrych** [reply]
Sorry, I meant a mechanism which would allow to migrate instances from one definition to another. So if the process changes, and I have a new definition for it - the long-running instances would have a possibility to be updated, to follow the new process.

Different engines handle it differently, but most of them have a possibility to somehow update the process instance without stopping it.

**Comment by ralphhanna** [comment]
Currently, model changes do not apply to started instances.

To support model changes to already started instances:. few options are possible to specify this requirement.

1. A `cli` command; to update instances of a particular `model` with revised bpmn source. 
2. System wide: in the `configuration.ts`, set an option to apply any model changes to running instances.

The first is safer, give you more control. Also, you specify change point, If instance passed certain point should not be modified.

The call to `cli` will be like this:
 Update Instances with new Model ( modelName, changeNode)
 this will be applied to `active` Instances of the `model` and Do not have the specified `changeNode` started

This can be combined to `restart` capability as well, by rewinding instances with the revised `model`

Your feedback please

**Reply by Dmytrych** [reply]
Thank you for the suggestions! I think I will go with the `cli` option then.
